### PR TITLE
opt: Avoid typed system adapter resolution on each typed actor start

### PR DESCRIPTION
**Pull request by johanandren**

### CODE REVIEW

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorSystemAdapter.scala [hunk 1]
```diff
@@ -129,15 +128,32 @@ import akka.annotation.InternalApi
...
-  def apply(system: classic.ActorSystem): ActorSystem[Nothing] = AdapterExtension(system).adapter
+  def apply(system: classic.ActorSystem): ActorSystem[Nothing] = {
+    system match {
+      case system: ActorSystemImpl =>
+        // Optimization to cut out going through adapter lookup if possible
+        system.typedSystem match {
+          case OptionVal.Some(typedSystem: ActorSystem[_]) => typedSystem
+          case _ =>
+            val typedSystem = AdapterExtension(system).adapter
+            system.typedSystem = OptionVal.Some(typedSystem)
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorSystemAdapter.scala:139]: Wonder if something is wrong here? Here it sets it to `ActorSystemAdapter`, but above it looks for `: ActorSystem[_]`?

File: akka-actor/src/main/scala/akka/actor/ActorSystem.scala [hunk 2]
```diff
@@ -820,6 +820,12 @@ private[akka] class ActorSystemImpl(
 
   private val _dynamicAccess: DynamicAccess = createDynamicAccess()
 
+  /**
+   * Optimistic optimization: Typed actors will _all_ pick up the system adapter, going through
+   * extension infra only once per thread
+   */
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/ActorSystem.scala:826]: Make the comment more clear about what type the `AnyRef` has. Also, I find the "once per thread" misleading, reading it like a thread local.

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorSystemAdapter.scala (hunk 1 above)
Comment (johanandren) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorSystemAdapter.scala:139]: `ActorSystemAdapter` is the (adapted) typed system. Would be fine to be explicit about it and match on the adapter type though.

Comment (patriknw) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorSystemAdapter.scala:139]: ah, ofc

**Review by patriknw — APPROVED** [review]
LGTM

**Review by franciscolopezsancho — APPROVED** [review]
LGTM

**Pull request merged**